db/migrations/20190711091453_create_flip_table.sql

```diff
@@ -135,256 +166,657 @@ COMMENT ON FUNCTION flip_bid_time_created
+)
+INSERT
+INTO maker.flip (bid_id, address_id, block_number, guy, tic, "end", lot, bid, usr, gal, tab, updated, created)
+VALUES (new_diff.bid_id,
+        new_diff.address_id,
+        (SELECT block_number FROM diff_block),
+        new_diff.guy,
+        flip_bid_tic_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_end_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_lot_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_bid_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_usr_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_gal_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        flip_bid_tab_before_block(new_diff.bid_id, new_diff.address_id, new_diff.header_id),
+        (SELECT api.epoch_to_datetime(block_timestamp) FROM diff_block),
+        flip_bid_time_created(new_diff.address_id, new_diff.bid_id))
+ON CONFLICT (block_number, bid_id, address_id) DO UPDATE SET guy = new_diff.guy;
 $$
-BEGIN
-    WITH diff_block AS (
-        SELECT block_number, block_timestamp
+    LANGUAGE sql;
+-- +goose StatementEnd
+
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION maker.update_flip_guys_until_next_diff(new_diff maker.flip_bid_guy) RETURNS VOID
+AS
+$$
+DECLARE
+    diff_block_number   BIGINT := (
+        SELECT block_number
         FROM public.headers
-        WHERE id = NEW.header_id
-    )
-    INSERT
-    INTO maker.flip (bid_id, address_id, block_number, guy, tic, "end", lot, bid, gal, tab, updated, created)
-    VALUES (NEW.bid_id, NEW.address_id, (SELECT block_number FROM diff_block), NEW.guy,
-            get_latest_flip_bid_tic(NEW.bid_id),
-            get_latest_flip_bid_end(NEW.bid_id),
-            get_latest_flip_bid_lot(NEW.bid_id),
-            get_latest_flip_bid_bid(NEW.bid_id),
-            get_latest_flip_bid_gal(NEW.bid_id),
-            get_latest_flip_bid_tab(NEW.bid_id),
-            (SELECT api.epoch_to_datetime(block_timestamp) FROM diff_block),
-            flip_bid_time_created(NEW.address_id, NEW.bid_id))
-    ON CONFLICT (block_number, bid_id, address_id) DO UPDATE SET guy = NEW.guy;
-    RETURN NEW;
+        WHERE id = new_diff.header_id);
+    next_guy_diff_block BIGINT := (
+        SELECT MIN(block_number)
+        FROM maker.flip_bid_guy
+                 LEFT JOIN public.headers ON flip_bid_guy.header_id = headers.id
+        WHERE flip_bid_guy.bid_id = new_diff.bid_id
+          AND flip_bid_guy.address_id = new_diff.address_id
+          AND block_number > diff_block_number);
+BEGIN
+    UPDATE maker.flip
+    SET guy = new_diff.guy
+    WHERE flip.bid_id = new_diff.bid_id
+      AND flip.address_id = new_diff.address_id
+      AND flip.block_number >= diff_block_number
+      AND (next_guy_diff_block IS NULL
+        OR flip.block_number < next_guy_diff_block);
```

Comment:
I did some reading up on composite indexes, and it looks like in this case Postgres will use the primary key index, since the columns we're filtering on are included in the index. Definitely a good idea to keep indexes in mind moving forward though!